File: seoProject/structureDataCheck.py
<<<<<<< HEAD
import json
import requests

# Step 1: Load the Schema.org JSON-LD vocabulary
def load_schema_vocabulary(url):
    try:
        response = requests.get(url)
        schema_data = response.json()
        return schema_data
    except Exception as e:
        print(f"Error loading Schema.org vocabulary: {e}")
        return None

# Step 2: Extract types and properties from the Schema.org vocabulary
def extract_schema_types_and_properties(schema_data):
    types_and_properties = {}

    # Iterate through the @graph part of the vocabulary, which contains types and properties
    for item in schema_data["@graph"]:
        if "@type" in item and item["@type"] == "Class":
            # Extract valid types
            types_and_properties[item["@id"]] = {
                "type": "Class",
                "properties": set()
            }

        if "@type" in item and item["@type"] == "Property":
            # Extract valid properties
            domain = item.get("domainIncludes", [])
            for domain_type in domain:
                if domain_type not in types_and_properties:
                    types_and_properties[domain_type] = {
                        "type": "Class",
                        "properties": set()
                    }
                types_and_properties[domain_type]["properties"].add(item["@id"])

    return types_and_properties

# Step 3: Validate the structured data
def validate_structured_data(structured_data, types_and_properties):
    issues = []

    # Check if @graph exists, if not treat the structured data as a single object
    if "@graph" in structured_data:
        items_to_validate = structured_data["@graph"]
    else:
        items_to_validate = [structured_data]

    # Validate each item in the graph (or the single root item)
    for item in items_to_validate:
        if "@type" not in item:
            issues.append("Missing '@type' key")
            continue
        
        item_type = item["@type"]
        
        if item_type not in types_and_properties:
            issues.append(f"Invalid type: '{item_type}'")
            continue

        properties = item.keys()
        valid_properties = types_and_properties.get(item_type, [])
        
        for prop in properties:
            if prop not in valid_properties:
                issues.append(f"Invalid property: '{prop}' for type '{item_type}'")

    return issues


# Step 4: Main function to run the validation
def main():
    schema_url = "https://schema.org/version/latest/schemaorg-current-https.jsonld"
    structured_data = """
    {
    "@context": "https://schema.org",
    "@type": "NewsArticle",
    "headline": "Racial and religious harmony scores rise in Singapore: IPS study",
    "name": "Racial and religious harmony scores rise in Singapore: IPS study",
    "description": "Researchers polled a representative sample of 4,000 Singaporeans and permanent residents aged 18 and above. Read more at straitstimes.com.",
    "image": {
        "@type": "ImageObject",
        "representativeOfPage": "True",
        "url": "https://cassette.sphdigital.com.sg/image/straitstimes/45d54436a1677b8a42fb080cd5a8c16bc77f7818fefeac42108cd7975a379103"
    },
    "datePublished": "2025-02-03T16:00:00+08:00",
    "dateModified": "2025-02-03T23:00:39+08:00",
    "isAccessibleForFree": "True",
    "author": [{
        "@type": "Person",
        "name": "Tham Yuen-C",
        "jobTitle": "Senior Political Correspondent",
        "url": "https://www.straitstimes.com/authors/tham-yuen-c"
    }],
    "publisher": {
        "@type": "Organization",
        "name": "The Straits Times"
    },
    "mainEntityOfPage": "https://www.straitstimes.com/singapore/racial-and-religious-harmony-scores-rise-in-spore-study"
    }

    """
    
    # Parse the structured data
    structured_data = json.loads(structured_data)

    # Load Schema.org vocabulary
    schema_data = load_schema_vocabulary(schema_url)
    if not schema_data:
        return

    # Extract types and properties from Schema.org
    types_and_properties = extract_schema_types_and_properties(schema_data)

    # Validate the structured data
    issues = validate_structured_data(structured_data, types_and_properties)

    # Output issues
    if issues:
        print("Validation Issues:")
        for issue in issues:
            print(f"- {issue}")
    else:
        print("No validation issues found!")

if __name__ == "__main__":
    main()
=======
import requests
from bs4 import BeautifulSoup
from jsonschema import validate, ValidationError
import json
from datetime import datetime

# Function to fetch the HTML content of a page
def fetch_html(url):
    try:
        response = requests.get(url)
        response.raise_for_status()
        return response.text
    except requests.exceptions.RequestException as e:
        print(f"Error fetching the URL: {e}")
        return None

# Function to fetch the Schema.org definition for a specific type
def fetch_schema_definition(schema_type="Thing"):
    url = "https://schema.org/version/latest/schemaorg-current-https.jsonld"
    try:
        response = requests.get(url)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error fetching schema definition: {e}")
        return None

# Function to validate the JSON-LD structured data against Schema.org definitions
def validate_jsonld(jsonld_data):
    results = {}
    try:
        schema_type = jsonld_data.get("@type", "Thing")
        schema = fetch_schema_definition(schema_type)
        if not schema:
            results["valid"] = False
            results["error"] = f"Schema definition not found for {schema_type}"
            results["details"] = None
            return results

        try:
            validate(instance=jsonld_data, schema=schema)
            results["valid"] = True
            results["error"] = None
            results["details"] = f"JSON-LD block is valid according to Schema.org definition for type '{schema_type}'."
        except ValidationError as e:
            results["valid"] = False
            results["error"] = e.message
            results["details"] = {
                "failed_validator": e.validator,
                "validator_value": e.validator_value,
                "instance_path": list(e.path),
                "schema_path": list(e.schema_path)
            }
    except json.JSONDecodeError:
        results["valid"] = False
        results["error"] = "Invalid JSON-LD format"
        results["details"] = None

    return results

# Function to extract JSON-LD structured data from HTML
def extract_jsonld_from_html(html_content):
    soup = BeautifulSoup(html_content, 'html.parser')
    scripts = soup.find_all('script', {'type': 'application/ld+json'})
    jsonld_blocks = []
    for script in scripts:
        if script.string:
            try:
                data = json.loads(script.string)
                if isinstance(data, list):
                    jsonld_blocks.extend(data)
                else:
                    jsonld_blocks.append(data)
            except json.JSONDecodeError as e:
                print("Error decoding JSON-LD:", e)
    
    # Debugging output to inspect extracted JSON-LD blocks and their types
    print("\n--- Extracted JSON-LD Blocks ---")
    for idx, block in enumerate(jsonld_blocks):
        print(f"Block #{idx + 1}: {json.dumps(block, indent=2)}")
        print(f"Type: {block.get('@type', 'No type specified')}")
    
    return jsonld_blocks

# Main function to perform the test
def test_structured_data_validation(url):
    # Step 1: Fetch HTML content from the URL
    html_content = fetch_html(url)
    if not html_content:
        print("Failed to fetch HTML content.")
        return

    # Step 2: Extract JSON-LD structured data
    jsonld_blocks = extract_jsonld_from_html(html_content)
    if not jsonld_blocks:
        print("No JSON-LD structured data found.")
        return

    # Initialize summary data
    valid_items = {}
    invalid_items = {}
    total_valid = 0
    total_invalid = 0

    # Step 3: Validate each JSON-LD block
    for idx, block in enumerate(jsonld_blocks):
        result = validate_jsonld(block)
        schema_type = block.get("@type", "Thing")
        if result["valid"]:
            if schema_type not in valid_items:
                valid_items[schema_type] = 0
            valid_items[schema_type] += 1
            total_valid += 1
        else:
            if schema_type not in invalid_items:
                invalid_items[schema_type] = 0
            invalid_items[schema_type] += 1
            total_invalid += 1

    # Step 4: Output summary
    print("\n--- Structured Data Validation Summary ---")
    print(f"Crawled successfully on {datetime.now().strftime('%b %d, %Y, %I:%M:%S %p')}")
    print(f"Total detected structured data: {len(jsonld_blocks)} items")
    
    # Display valid items summary
    if valid_items:
        print("\nValid items:")
        for item_type, count in valid_items.items():
            print(f"  {item_type}: {count} valid item(s) detected")
    
    # Display invalid items summary
    if invalid_items:
        print("\nInvalid items:")
        for item_type, count in invalid_items.items():
            print(f"  {item_type}: {count} invalid item(s) detected")
    
    # Display overall valid and invalid counts
    print(f"\nTotal valid items detected: {total_valid}")
    print(f"Total invalid items detected: {total_invalid}")
    
    # Display Google Rich Results eligibility
    if total_valid > 0:
        print("\nValid items are eligible for Google Search's rich results. Learn more.")
    else:
        print("\nNo valid items detected for rich results.")

if __name__ == "__main__":
    url = "https://www.channelnewsasia.com/commentary/education-master-degrees-qualifications-useless-tuition-fees-4802721"  # Example URL
    test_structured_data_validation(url)
>>>>>>> e47443c8824bdd271d25061d498d4ea2afd0b41f

File: backend/seoProject/structureDataCheck.mjs
// //v7
// import jsonld from 'jsonld';
// import fetch from 'node-fetch';
// import * as cheerio from 'cheerio';

// // Function to fetch and validate JSON-LD from a webpage
// async function validateJsonLdFromUrl(url) {
//     try {
//         // Fetch the webpage content
//         const response = await fetch(url);
        
//         // Check if the response was successful (status 200)
//         if (!response.ok) {
//             console.error(`🚨 Error: Failed to fetch the URL. Status: ${response.status} ${response.statusText}`);
//             return;
//         }

//         const html = await response.text();

//         // Load HTML content with cheerio
//         const $ = cheerio.load(html);

//         // Extract JSON-LD scripts
//         const jsonLdScripts = [];
//         $('script[type="application/ld+json"]').each((i, element) => {
//             const jsonLd = $(element).html();
//             console.log(`🔍 Extracted JSON-LD from Script ${i + 1}:`);
//             console.log(jsonLd); // Logging the raw JSON-LD content
//             try {
//                 const parsedJsonLd = JSON.parse(jsonLd);
//                 jsonLdScripts.push(parsedJsonLd);
//             } catch (error) {
//                 console.error(`❌ Error parsing JSON-LD script ${i + 1}:`, error.message);
//             }
//         });

//         if (jsonLdScripts.length === 0) {
//             console.log('⚠️ No JSON-LD data found on the page.');
//             return;
//         }

//         // Track the valid items and their types
//         let validItemCount = 0;
//         const validItems = [];

//         // Loop through each JSON-LD script and extract the type and properties
//         for (let i = 0; i < jsonLdScripts.length; i++) {
//             const jsonLdData = jsonLdScripts[i];
//             let scriptValidItems = [];

//             try {
//                 // Check if the JSON-LD contains a '@graph' key
//                 if (jsonLdData['@graph']) {
//                     const graph = jsonLdData['@graph'];

//                     graph.forEach(item => {
//                         // Check for a valid @type for each item in the @graph
//                         const jsonLdType = item['@type'];
//                         if (jsonLdType) {
//                             scriptValidItems.push(jsonLdType);
//                             // Log each item type found
//                             console.log(`🔑 Found valid item in @graph: ${jsonLdType}`);
                            
//                             // Log the elements of the valid item
//                             console.log(`  Details for ${jsonLdType}:`);
//                             Object.keys(item).forEach(key => {
//                                 if (key !== '@type') { // Exclude '@type' key as we already logged it
//                                     console.log(`    ${key}: ${JSON.stringify(item[key], null, 2)}`);
//                                 }
//                             });
//                         }
//                     });

//                     validItemCount += scriptValidItems.length;

//                     // Log the number of valid structured items
//                     console.log(`Valid structured items found in @graph: ${scriptValidItems.length}`);
//                 } else {
//                     console.log(`⚠️ JSON-LD Script ${i + 1} does not have a valid '@graph' field.`);
//                 }

//                 // Compact JSON-LD to validate context
//                 const compacted = await jsonld.compact(jsonLdData, {});
//                 console.log(`✅ JSON-LD script ${i + 1} is valid:\n`, JSON.stringify(compacted, null, 2));

//             } catch (err) {
//                 console.error(`❌ JSON-LD validation failed for script ${i + 1}:`, err.message);
//             }

//             // Store valid items for final summary
//             validItems.push({
//                 scriptIndex: i + 1,
//                 validItemTypes: scriptValidItems
//             });
//         }

//         // Log the total number of valid structured items (based on @type in all scripts)
//         console.log(`\n📊 Total valid items: ${validItemCount}`);

//         // Log a summary of valid items for each script
//         validItems.forEach((item, index) => {
//             console.log(`\n🏷️ Script ${item.scriptIndex}:`);
//             item.validItemTypes.forEach(type => {
//                 console.log(`- ${type}`);
//             });
//         });

//     } catch (error) {
//         console.error('🚨 Error fetching or validating structured data:', error.message);
//     }
// }

// // Example usage
// const websiteUrl = 'https://www.semrush.com/kb/995-what-is-semrush'; // Replace with the target website URL
// await validateJsonLdFromUrl(websiteUrl);


//v8 returning the result instead of logging them 
// import jsonld from 'jsonld';
// import fetch from 'node-fetch';
// import * as cheerio from 'cheerio';

// // Function to fetch and validate JSON- LD from a webpage
// async function validateJsonLdFromUrl(url) {
//     try {
//         // Fetch the webpage content
//         const response = await fetch(url);
        
//         // Check if the response was successful (status 200)
//         if (!response.ok) {
//             console.error(`🚨 Error: Failed to fetch the URL. Status: ${response.status} ${response.statusText}`);
//             return;
//         }

//         const html = await response.text();

//         // Load HTML content with cheerio
//         const $ = cheerio.load(html);

//         // Extract JSON-LD scripts
//         const jsonLdScripts = [];
//         $('script[type="application/ld+json"]').each((i, element) => {
//             const jsonLd = $(element).html();
//             console.log(`🔍 Extracted JSON-LD from Script ${i + 1}:`);
//             console.log(jsonLd); // Logging the raw JSON-LD content
//             try {
//                 const parsedJsonLd = JSON.parse(jsonLd);
//                 jsonLdScripts.push(parsedJsonLd);
//             } catch (error) {
//                 console.error(`❌ Error parsing JSON-LD script ${i + 1}:`, error.message);
//             }
//         });

//         if (jsonLdScripts.length === 0) {
//             console.log('⚠️ No JSON-LD data found on the page.');
//             return;
//         }

//         // Track the valid items and their types
//         let validItemCount = 0;
//         const validItems = [];
//         const validItemTypes = new Set(); // To track unique valid item types

//         // Loop through each JSON-LD script and extract the type and properties
//         for (let i = 0; i < jsonLdScripts.length; i++) {
//             const jsonLdData = jsonLdScripts[i];
//             let scriptValidItems = [];

//             try {
//                 // Check if the JSON-LD contains a '@graph' key
//                 if (jsonLdData['@graph']) {
//                     const graph = jsonLdData['@graph'];

//                     graph.forEach(item => {
//                         // Check for a valid @type for each item in the @graph
//                         const jsonLdType = item['@type'];
//                         if (jsonLdType) {
//                             scriptValidItems.push({
//                                 type: jsonLdType,
//                                 details: item
//                             });
//                             validItemTypes.add(jsonLdType); // Add to set of valid item types
//                             // Log each item type found
//                             console.log(`🔑 Found valid item in @graph: ${jsonLdType}`);
                            
//                             // Log the elements of the valid item
//                             console.log(`  Details for ${jsonLdType}:`);
//                             Object.keys(item).forEach(key => {
//                                 if (key !== '@type') { // Exclude '@type' key as we already logged it
//                                     console.log(`    ${key}: ${JSON.stringify(item[key], null, 2)}`);
//                                 }
//                             });
//                         }
//                     });

//                     validItemCount += scriptValidItems.length;

//                     // Log the number of valid structured items
//                     console.log(`Valid structured items found in @graph: ${scriptValidItems.length}`);
//                 } else {
//                     console.log(`⚠️ JSON-LD Script ${i + 1} does not have a valid '@graph' field.`);
//                 }

//                 // Compact JSON-LD to validate context
//                 const compacted = await jsonld.compact(jsonLdData, {});
//                 console.log(`✅ JSON-LD script ${i + 1} is valid:\n`, JSON.stringify(compacted, null, 2));

//             } catch (err) {
//                 console.error(`❌ JSON-LD validation failed for script ${i + 1}:`, err.message);
//             }

//             // Store valid items for final summary
//             validItems.push({
//                 scriptIndex: i + 1,
//                 validItems: scriptValidItems
//             });
//         }

//         // Return the results: valid items, number of valid items, valid item types, and the elements in the valid items
//         return {
//             totalValidItems: validItemCount,
//             validItems,
//             validItemTypes: Array.from(validItemTypes) // Convert Set to Array for easy access
//         };

//     } catch (error) {
//         console.error('🚨 Error fetching or validating structured data:', error.message);
//         return {
//             totalValidItems: 0,
//             validItems: [],
//             validItemTypes: []
//         };
//     }
// }

// // Example usage
// const websiteUrl = 'https://www.channelnewsasia.com/commentary/education-master-degrees-qualifications-useless-tuition-fees-4802721'; // Replace with the target website URL
// const result = await validateJsonLdFromUrl(websiteUrl);
// console.log(`\n📊 Total valid items: ${result.totalValidItems}`);
// console.log(`\n📋 Valid Item Types: ${result.validItemTypes.join(', ')}`);
// result.validItems.forEach(item => {
//     console.log(`\n🏷️ Script ${item.scriptIndex}:`);
//     item.validItems.forEach(validItem => {
//         console.log(`- ${validItem.type}`);
//         console.log(`  Details: ${JSON.stringify(validItem.details, null, 2)}`);
//     });
// });


//v9 turning into a script for running via a call
import jsonld from 'jsonld';
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

// Function to fetch and validate JSON-LD from a webpage
async function validateJsonLdFromUrl(url) {
    try {
        // Fetch the webpage content
        const response = await fetch(url);
        
        // Check if the response was successful (status 200)
        if (!response.ok) {
            throw new Error(`Failed to fetch the URL. Status: ${response.status} ${response.statusText}`);
        }

        const html = await response.text();

        // Load HTML content with cheerio
        const $ = cheerio.load(html);

        // Extract JSON-LD scripts
        const jsonLdScripts = [];
        $('script[type="application/ld+json"]').each((i, element) => {
            const jsonLd = $(element).html();
            try {
                const parsedJsonLd = JSON.parse(jsonLd);
                jsonLdScripts.push(parsedJsonLd);
            } catch (error) {
                throw new Error(`Error parsing JSON-LD script ${i + 1}: ${error.message}`);
            }
        });

        if (jsonLdScripts.length === 0) {
            throw new Error('No JSON-LD data found on the page.');
        }

        // Track the valid items and their types
        let validItemCount = 0;
        const validItems = [];
        const validItemTypes = new Set(); // To track unique valid item types

        // Loop through each JSON-LD script and extract the type and properties
        for (let i = 0; i < jsonLdScripts.length; i++) {
            const jsonLdData = jsonLdScripts[i];
            let scriptValidItems = [];

            try {
                // Check if the JSON-LD contains a '@graph' key
                if (jsonLdData['@graph']) {
                    const graph = jsonLdData['@graph'];

                    graph.forEach(item => {
                        // Check for a valid @type for each item in the @graph
                        const jsonLdType = item['@type'];
                        if (jsonLdType) {
                            scriptValidItems.push({
                                type: jsonLdType,
                                details: item
                            });
                            validItemTypes.add(jsonLdType); // Add to set of valid item types
                        }
                    });

                    validItemCount += scriptValidItems.length;
                }

                // Compact JSON-LD to validate context
                await jsonld.compact(jsonLdData, {});

            } catch (err) {
                throw new Error(`JSON-LD validation failed for script ${i + 1}: ${err.message}`);
            }

            // Store valid items for final summary
            validItems.push({
                scriptIndex: i + 1,
                validItems: scriptValidItems
            });
        }

        // Return the structured result
        return {
            totalValidItems: validItemCount,
            validItems,
            validItemTypes: Array.from(validItemTypes), // Convert Set to Array for easy access
        };

    } catch (error) {
        // Return a structured error result if an error occurred
        return {
            error: error.message,
        };
    }
}

//Example usage in another module
async function main() {
    const websiteUrl = 'https://www.screamingfrog.co.uk/generate-json-ld-schema-at-scale/'; // Replace with the target website URL
    const result = await validateJsonLdFromUrl(websiteUrl);

    // Example of structured result for external use
    console.log(JSON.stringify(result, null, 2)); // Will be in a structured format
}

main().catch(err => console.error("Error in JSON-LD validation:", err));


// Example usage
// const websiteUrl = 'https://firstpagesage.com/seo-blog/top-seo-tools-comparison/'; // Replace with the target website URL
// const result = await validateJsonLdFromUrl(websiteUrl);
// console.log(`\n📊 Total valid items: ${result.totalValidItems}`);
// console.log(`\n📋 Valid Item Types: ${result.validItemTypes.join(', ')}`);
// result.validItems.forEach(item => {
//     console.log(`\n🏷️ Script ${item.scriptIndex}:`);
//     item.validItems.forEach(validItem => {
//         console.log(`- ${validItem.type}`);
//         console.log(`  Details: ${JSON.stringify(validItem.details, null, 2)}`);
//     });
// });

